### ISSUE TITLE
`curl` on _OpenWrt_ is not printing **certificate info** like it does on _macOS_

### ISSUE THREAD

**Issue by Qlub53** [curl/curl#9674]
# Problem Statement

`curl` on _OpenWrt_ is not printing **certificate info** like it does on _macOS_.

--------------------
### I did this

```sh
curl --verbose https://raw.githubusercontent.com
```

output on OpenWrt skips much information and begins with handshakes
```
* TLSv1.3 (OUT), TLS handshake, Client hello (1):
* TLSv1.3 (IN), TLS handshake, Server hello (2):
* TLSv1.3 (IN), TLS handshake, Encrypted Extensions (8):
* TLSv1.3 (IN), TLS handshake, Certificate (11):
* TLSv1.3 (OUT), TLS alert, unknown CA (560):
* SSL certificate problem: self signed certificate
curl: (60) SSL certificate problem: self signed certificate
More details here: https://curl.se/docs/sslcerts.html

curl failed to verify the legitimacy of the server and therefore could not
establish a secure connection to it. To learn more about this situation and
how to fix it, please visit the web page mentioned above.
```

output on macOS covers more details with IP & port, protocol, certificate and handshakes, etc
```
*   Trying 198.18.0.35:443...
* Connected to raw.githubusercontent.com (198.18.0.35) port 443 (#0)
* ALPN, offering h2
* ALPN, offering http/1.1
* successfully set certificate verify locations:
*  CAfile: /etc/ssl/cert.pem
*  CApath: none
* (304) (OUT), TLS handshake, Client hello (1):
* (304) (IN), TLS handshake, Server hello (2):
* (304) (IN), TLS handshake, Unknown (8):
* (304) (IN), TLS handshake, Certificate (11):
* (304) (IN), TLS handshake, CERT verify (15):
* (304) (IN), TLS handshake, Finished (20):
* (304) (OUT), TLS handshake, Finished (20):

...(and more omitted)
```

### I expected the following

I expected `curl` printing where it loaded the certificate from, like on macOS, e.g.
```
* successfully set certificate verify locations:
*  CAfile: /etc/ssl/cert.pem
*  CApath: none
```
before printing handshakes
```
* TLSv1.3 (OUT), TLS handshake, Client hello (1):
* TLSv1.3 (IN), TLS handshake, Server hello (2):
* TLSv1.3 (IN), TLS handshake, Encrypted Extensions (8):
* TLSv1.3 (IN), TLS handshake, Certificate (11):
* TLSv1.3 (OUT), TLS alert, unknown CA (560):
```
so that I could proceed with debugging my certificate issue.

### curl/libcurl version

on OpenWrt
```sh
curl 7.79.1 (aarch64-openwrt-linux-gnu) libcurl/7.79.1 OpenSSL/1.1.1l nghttp2/1.44.0 OpenLDAP/2.4.58
Release-Date: 2021-09-22
Protocols: file ftp ftps http https ldap ldaps mqtt tftp
Features: alt-svc AsynchDNS HSTS HTTP2 HTTPS-proxy IPv6 Largefile SSL TLS-SRP UnixSockets
```

on macOS
```sh
curl 7.79.1 (x86_64-apple-darwin21.0) libcurl/7.79.1 (SecureTransport) LibreSSL/3.3.6 zlib/1.2.11 nghttp2/1.45.1
Release-Date: 2021-09-22
Protocols: dict file ftp ftps gopher gophers http https imap imaps ldap ldaps mqtt pop3 pop3s rtsp smb smbs smtp smtps telnet tftp 
Features: alt-svc AsynchDNS GSS-API HSTS HTTP2 HTTPS-proxy IPv6 Kerberos Largefile libz MultiSSL NTLM NTLM_WB SPNEGO SSL UnixSockets
```

### operating system

on OpenWrt
```
Linux ImmortalWrt 5.4.152 #0 SMP PREEMPT aarch64 GNU/Linux
```
on macOS
```
Darwin MacBook-PRO.local 21.6.0 Darwin Kernel Version 21.6.0; root:xnu-****.***.**~1/RELEASE_ARM64_T6000 arm64
```

**Comment by dfandrich** [comment]
Was curl on OpenWrt built with --disable-verbose perhaps? That wouldn't surprise me, given that it's designed for an embedded application.

**Comment by Qlub53** [comment]
> Was curl on OpenWrt built with --disable-verbose perhaps? That wouldn't surprise me, given that it's designed for an embedded application.
> 

That's a good point. But I'm not sure how to check if it was built with `--disable-verbose`.

`--disable-verbose` "[eliminates debugging strings and error code strings](https://github.com/curl/curl/blob/master/docs/INSTALL.md#:~:text=eliminates%20debugging%20strings%20and%20error%20code%20strings)". Sounds like it would disable `curl --verbose` altogether, i.e. no output at all when running `curl --verbose`?
Yet, my log shows it's only missing **_some_** output but not all.

**Comment by jay** [comment]
Yes, it looks like libcurl was built with [CURL_DISABLE_VERBOSE_STRINGS](https://github.com/curl/curl/blob/93d092867f0f2c78571983040ef75e078ee1a4c4/docs/CURL-DISABLE.md#curl_disable_verbose_strings) (build: --disable-verbose), which [disables infof()](https://github.com/curl/curl/blob/93d092867f0f2c78571983040ef75e078ee1a4c4/lib/sendf.h#L32-L46) that is used in [many places](https://github.com/curl/curl/blob/93d092867f0f2c78571983040ef75e078ee1a4c4/lib/vtls/openssl.c#L3117-L3118) to output verbose text. infof() normally would call Curl_debug() which is not disabled in any case.

Curl_debug is sometimes called directly by different code in libcurl. For openssl there is a tracing callback which will call Curl_debug if [CURLOPT_VERBOSE](https://curl.se/libcurl/c/CURLOPT_VERBOSE.html) _and_ [CURLOPT_DEBUGFUNCTION](https://curl.se/libcurl/c/CURLOPT_DEBUGFUNCTION.html) were set by the application (the application in this case is the curl tool which has enabled both).

Other calls to Curl_debug in libcurl are not guarded, they just call it and it has its own check that CURLOPT_VERBOSE (ie data->set.verbose) was set. What this means is if libcurl was built with CURL_DISABLE_VERBOSE_STRINGS, even if the application does not set CURLOPT_DEBUGFUNCTION (ie data->set.fdebug) then libcurl can still output some data via Curl_debug when CURLOPT_VERBOSE is set.

I don't think Curl_debug can be a no-op when CURL_DISABLE_VERBOSE_STRINGS is defined because there may be user expectation around CURLOPT_DEBUGFUNCTION. For example the tool option --trace uses it. AFAICT there's no programmatic way for the application to check if CURL_DISABLE_VERBOSE_STRINGS was set at build time.

https://github.com/curl/curl/blob/93d092867f0f2c78571983040ef75e078ee1a4c4/lib/sendf.c#L718-L743

**Comment by bagder** [comment]
The [openwrt / package / network / utils / curl / Makefile](https://git.openwrt.org/?p=openwrt/openwrt.git;a=blob;f=package/network/utils/curl/Makefile;h=950044a226e9c85bd70e2c44b5115a472143268b;hb=558680012dd79de0289aeeb74a72163916cd9f33) has `--disable-manual` hardcoded in there.

**Comment by Qlub53** [comment]
> > Was curl on OpenWrt built with --disable-verbose perhaps? That wouldn't surprise me, given that it's designed for an embedded application.
> > 
> 
> That's a good point. But I'm not sure how to check if it was built with `--disable-verbose`.
> 
> `--disable-verbose` "[eliminates debugging strings and error code strings](https://github.com/curl/curl/blob/master/docs/INSTALL.md#:~:text=eliminates%20debugging%20strings%20and%20error%20code%20strings)". Sounds like it would disable `curl --verbose` altogether, i.e. no output at all when running `curl --verbose`?
> Yet, my log shows it's only missing **_some_** output but not all.

Okay, given that `libcurl` on openwrt is built with `--disable-verbose`, imo it's supposed to 
> disable `curl --verbose` altogether

But we still see handshakes were printed by `curl --verbose`. Aren't handshakes info verbose, too?

Please correct me if this is not the case.

**Comment by dfandrich** [comment]
The intent of --disable-verbose is to reduce the size of the binary in
applications where the verbose messages won't be seen, anyway, by eliminating
storing all the message strings. Messages coming from other libraries (like
OpenSSL) aren't affected by this option since the strings are stored in the
other libraries. Eliminating those as well would safe a few bytes per message,
but not nearly as much as curl messages. Also, some of the messages you're
seeing might be error messages which aren't removed.

**Comment by bagder** [comment]
This issue asked about the differences. They have been explained in detail now. This is not a bug.
